### DISCUSSION TITLE
k3s rootless - two step install

### DISCUSSION THREAD

**Discussion by vsoch** [k3s-io/k3s#7615] [Q&A] [answered]
Hi! I'm trying to prepare a set of VMs with k3s, and have everything that I need to (later) allow a user to login and setup rootless. With the current suggested workflow, namely calling the install script in one of two ways:

```bash
# For the control plane
curl -sfL https://get.k3s.io | K3S_TOKEN=${secret_token} sh -

# For an agent
curl -sfL https://get.k3s.io | INSTALL_K3S_CHANNEL=testing K3S_URL=https://${login_node}:6443 K3S_TOKEN=${secret_token} sh -
```
This assumes that I can get sudo, so I can't have this be done by a user. So I've been trying to piece apart the logic of that script to separate the two needs (root vs user run) and I found both of these:

```
INSTALL_K3S_SKIP_ENABLE
INSTALL_K3S_SKIP_START
```
And that seems to successfully install the binaries, but then when I need to generate the kubeconfig for the user, e.g.,

```
cp /etc/rancher/k3s/k3s.yaml ~/.kube/config
```
that file isn't there. I could have the full startup script with some extra envars in the boot script, but we don't actually have hostnames yet and it's run for both control plane / agent nodes, so I can't be specific to choose which one to run. So I want to ask (and hopefully this is an easy answer!) how do I install this in a way that I can achieve the above - having the binaries / configs mostly ready to go on a system (prepared by root) and then a user just needs to setup their rootless service? Thank you!

**Comment by brandond** [comment]
Are you really talking about using k3s [rootless](https://docs.k3s.io/advanced#running-rootless-servers-experimental), or are you just talking about installing and running k3s normally (as root, via a normal systemd service) and allowing users to access it without any additional privilege?

I will note that if you give a non-root user admin access to Kubernetes you might as well just give them root as well; they can fully escalate to root by creating privileged pods that mount the host filesystem.

**Reply by vsoch** [reply]
The scenario I have is that users are going to spin up k3s for an interactive job, which will be able to have nodes as the control plane and agents. So I need to have most of the "requires sudo" stuff ready to go, and then a few commands that the user can issue when the job starts to create the service, get the kubeconfig, and interact with their ephemeral cluster. I feel like I'm close, but I'm running into these edge issues. Since posting, I've tried adding:

```
ExecStart=/usr/local/bin/k3s server --rootless --snapshotter=fuse-overlayfs --write-kubeconfig=~/.kube/config --write-kubeconfig-mode=600 --token=<some-hard-coded-token>
```
to the service file (which is generated just for the job) and the `--write-kubeconfig` doesn't seem to actually work, it generates a symbolic link to a file that doesn't exist. E.g.,:

```
$ ls -l ~/.kube/
total 4
drwxr-x---. 4 vsochat_gmail_com vsochat_gmail_com   35 May 26 20:42 cache
lrwxrwxrwx. 1 vsochat_gmail_com vsochat_gmail_com   14 May 26 20:39 k3s.yaml -> '~/.kube/config'
```
The only way I can get it working is to copy from the /etc/rancher directory (requires sudo, won't work in production) and then change ownership to the user. Ideally I could run a start command that would reliably generate the file in the path I've specified, and with the correct permission. When I do that manually, I can confirm my control plane looks okay (superficially):

```
$ kubectl --kubeconfig=$HOME/.kube/config get nodes
NAME             STATUS   ROLES                  AGE   VERSION
gffw-login-001   Ready    control-plane,master   57m   v1.26.4+k3s1
```
This is just for the control plane - I'm still fighting with starting the agent nodes under rootless. What I currently have is:

```
export INSTALL_K3S_CHANNEL=testing 
export K3S_URL=https://${login_node}:6443
export K3S_TOKEN=${secret_token}
# Note the -data-dir should default to this for the server
# k3s agent --rootless --debug --data-dir=~/.rancher/k3s/rootless
systemd-run --user -p Delegate=yes --tty k3s agent --rootless --debug --data-dir=~/.rancher/k3s/rootless
```
```console
Failed to create bus connection: No such file or directory
```
And I only added the system-d wrapper (which I don't think I should need to do) because when I didn't, it gave me an error that to run rootless with cgroups v2 I needed it (logical). So I'm not sure what to try there either. Thanks for your help!

**Reply by vsoch** [reply]
Forgot to answer this!

> I will note that if you give a non-root user admin access to Kubernetes you might as well just give them root as well; they can fully escalate to root by creating privileged pods that mount the host filesystem.

This is on a supercomputer. We absolutely cannot do that. We will have to understand and deal with the other issues as you mentioned later, right now this is all experimental and we are just trying to get it to run with rootless.

**Reply by vsoch** [reply]
For the agent, here is what I'm trying (and the output). It doesn't seem to be taking the `--server` flag I'm providing, unless the config server is different and should be local?

```bash
$ k3s agent --rootless --debug --data-dir=${HOME}/.rancher/k3s/rootless --token=${secret_token} --server=https://${login_address}:6443
```
```console
INFO[0000] Starting k3s agent v1.26.4+k3s1 (8d0255af)   
DEBU[0000] Running rootless parent                      
WARN[0000] Cannot set cgroup2 evacuation, make sure to run k3s as a systemd unit 
DEBU[0000] executing [nsenter -t 14811 -n -m -U --preserve-credentials ip tuntap add name tap0 mode tap] 
DEBU[0000] executing [nsenter -t 14811 -n -m -U --preserve-credentials ip link set tap0 up] 
DEBU[0000] sent tapfd=8 for tap0                        
DEBU[0000] received tapfd=8
Starting slirp              
INFO[0000] Starting k3s agent v1.26.4+k3s1 (8d0255af)   
DEBU[0000] Running rootless child                       
DEBU[0000] child: got msg from parent: {Stage:0 Message0:{} Message1:{StateDir: Network:{Dev: IP: Netmask:0 Gateway: DNS: MTU:0 Opaque:map[]} Port:{Opaque:map[]}}} 
INFO[0000] Starting k3s agent v1.26.4+k3s1 (8d0255af)   
DEBU[0000] Running rootless child                       
DEBU[0000] child: got msg from parent: {Stage:1 Message0:{} Message1:{StateDir:/tmp/rootless2224101104 Network:{Dev:tap0 IP:10.41.0.100 Netmask:16 Gateway:10.41.0.2 DNS:10.41.0.3 MTU:65520 Opaque:map[]} Port:{Opaque:map[builtin.readypipepath:/tmp/rootless2224101104/.bp-ready.pipe builtin.socketpath:/tmp/rootless2224101104/.bp.sock]}}} 
DEBU[0000] executing [ip link set lo up]                
DEBU[0000] executing [ip link set tap0 up]              
DEBU[0000] executing [ip link set dev tap0 mtu 65520]   
DEBU[0000] executing [ip addr add 10.41.0.100/16 dev tap0] 
DEBU[0000] executing [ip route add default via 10.41.0.2 dev tap0] 
INFO[0000] Starting k3s agent v1.26.4+k3s1 (8d0255af)   
DEBU[0000] Running rootless process                     
DEBU[0000] Mounting none /var/log tmpfs                 
DEBU[0000] Mounting /home/vsochat_gmail_com/.rancher/k3s/rootless/logs/var/log none bind 
DEBU[0000] Mounting none /var/lib/cni tmpfs             
DEBU[0000] Mounting /home/vsochat_gmail_com/.rancher/k3s/rootless/cni/var/lib/cni none bind 
DEBU[0000] Mounting none /var/lib/kubelet tmpfs         
DEBU[0000] Mounting /home/vsochat_gmail_com/.rancher/k3s/rootless/kubelet/var/lib/kubelet none bind 
DEBU[0000] Mounting none /etc/rancher tmpfs             
DEBU[0000] Mounting /home/vsochat_gmail_com/.rancher/k3s/rootless/etc/rancher/etc/rancher none bind 
INFO[0000] Adding server to load balancer k3s-agent-load-balancer: 10.10.0.6:6443 
INFO[0000] Running load balancer k3s-agent-load-balancer 127.0.0.1:6444 -> [10.10.0.6:6443] [default: 10.10.0.6:6443] 
WARN[0000] Cluster CA certificate is not trusted by the host CA bundle, but the token does not include a CA hash. Use the full token from the server's node-token file to enable Cluster CA validation. 
INFO[0000] Waiting to retrieve agent configuration; server is not ready: failed to retrieve configuration from server: https://127.0.0.1:6444/v1-k3s/config: 401 Unauthorized 
INFO[0005] Waiting to retrieve agent configuration; server is not ready: failed to retrieve configuration from server: https://127.0.0.1:6444/v1-k3s/config: 401 Unauthorized 
```
Sorry if I'm just being very stupid in some way I'm fairly new to k3s!

**Reply by brandond** [reply]
Please read through the rootless docs I linked to, and check out the example rootless systemd unit that it references. Also note a very important limitation that you appear to have missed:

> Multi-node rootless clusters, or multiple rootless k3s processes on the same node, are not currently supported. See [#6488](https://github.com/k3s-io/k3s/issues/6488#issuecomment-1314998091) for more details.

In plain english: you cannot register agents to a rootless server, and you cannot run rootless agents against a traditional rooted server.

**Reply by vsoch** [reply]
> Please read through the rootless docs I linked to, and check out the example rootless systemd unit that it references. Also note a very important limitation that you appear to have missed:

Yes thank you, I have been reading those docs for weeks, and I have that service file but it's starting the main server (and not an agent). I only was looking at one for the agent because it gave me in an error I should have one.

>  In plain english: you cannot register agents to a rootless server, and you cannot run rootless agents against a traditional rooted server.

Oh, I read that as "you can't run them on the same node" and not that it wouldn't work across nodes. In that issue, the comment seems to be trying to run rootless on one node, but I have several here, but you are saying that won't work (and why not?). It does feel like I'm close (but based on your wisdom I'm likely not going to get there) - when I try to use `node-token` for the token for the agent, it tells me:

```
ERRO[0008] token CA hash does not match the Cluster CA certificate hash: b1a1873410c51324d2ca1e01f72a9520bf4f4bc40fc27496195162120b9fca7e != ac0e954e04cf85534d5ac938e05e1ec1de5091f0efac4927f062a2584c9bfbfb
```

And then with the actual token (what we saw above). Is there any reason that address that is trying to hit localhost cannot hit the main control node? Can you help me better understand why this absolutely won't work?

**Reply by brandond** [reply]
It might help to start over from the beginning - what are your goals, and what are the restrictions? It sounds like your goal is to allow users to run K3s on multiple nodes, without the users having root on those nodes? I'm guessing you're also planning on automating all of this via a batch scheduling system of some sort? I'm not sure I can think of a good way to allow that, given the current limitations of rootless Kubernetes.

**Reply by brandond** [reply]
> Is there any reason that address that is trying to hit localhost cannot hit the main control node? Can you help me better understand why this absolutely won't work?

Did you click through where it says

>See https://github.com/k3s-io/k3s/issues/6488#issuecomment-1314998091 for more details.

There is a paragraph-long discussion of what problems I ran into last time I tried to get this working, most of it revolving around the way that rootless Kubernetes runs in a separate network namespace, and bridges everything through slirp4netns. None of the inter-node stuff works properly through that, for several reasons.

**Reply by vsoch** [reply]
> It sounds like your goal is to allow users to run K3s on multiple nodes, without the users having root on those nodes?

Yes that is correct. Right now I'm developing in a terraform deployed set of nodes that have Flux Framework installed (the job manager that allows us to do this).

> I'm guessing you're also planning on automating all of this via a batch scheduling system of some sort? I'm not sure I can think of a good way to allow that, given the current limitations of rootless Kubernetes.

I've already figured out that part - here is how we do that with Flux, and inside of a Kubernetes operator, albeit we allow sudo. https://github.com/flux-framework/flux-operator/tree/main/examples/nested/k3s/basic. To get that working I had to build a custom image that already had k3s installed since we don't want to wait during the start of the Flux Operator MiniCluster (this is an indexed job that creates a Flux cluster so it's expected to come up quickly, mostly just puling a container): https://github.com/rse-ops/flux-hpc/tree/main/k3s

What I'm doing now is translating this to rootless, so I've moved outside of Kubernetes and am using Flux installed on VMs via terraform, and then a combination of boot scripts and what is called a flux batch (creating a grouped set of commands under the same space owned by the user) to launch the cluster. Aside from the issue of getting the ~/.kube/config in the right place without sudo and this last step of starting the agent, at least superficially it seems I'm pretty close. Here is what I have so far if you are interested. https://github.com/converged-computing/flux-terraform-gcp/tree/test/rootless-k3s/examples/k3s

Maybe there is still something to try? The main issue seems to be it wants a configuration:

```
INFO[0000] Adding server to load balancer k3s-agent-load-balancer: 10.10.0.6:6443 
INFO[0000] Running load balancer k3s-agent-load-balancer 127.0.0.1:6444 -> [10.10.0.6:6443] [default: 10.10.0.6:6443] 
WARN[0000] Cluster CA certificate is not trusted by the host CA bundle, but the token does not include a CA hash. Use the full token from the server's node-token file to enable Cluster CA validation. 
INFO[0000] Waiting to retrieve agent configuration; server is not ready: failed to retrieve configuration from server: https://127.0.0.1:6444/v1-k3s/config: 401 Unauthorized 
```
And I'll go back and look at that issue again but I'm afraid I don't understand the components deeply enough to truly parse it.

**Reply by vsoch** [reply]
> There is a paragraph-long discussion of what problems I ran into last time I tried to get this working, 

I see the comment you linked to, but the context is running this on one node, which is different than what I'm doing because I have one node per control plane and each agent, and a full networking setup. Sorry if I'm missing the paragraph you are referencing! This is my case:

> There is also the larger issue of making this work across multiple hosts, which I think would have a bunch of challenges that I haven't even thought of yet.

**Reply by vsoch** [reply]
Could it be this? https://github.com/k3s-io/k3s/issues/5459#issuecomment-1103225394

```
]$ ping -M do -s 8972 $login_address
PING 10.10.0.6 (10.10.0.6) 8972(9000) bytes of data.
ping: local error: Message too long, mtu=1460
ping: local error: Message too long, mtu=1460
ping: local error: Message too long, mtu=1460
ping: local error: Message too long, mtu=1460
```

**Reply by brandond** [reply, marked as answer]
Rootless K3s runs fully isolated with its own pid/user/network namespaces. All instances of rootless k3s will use the same node IP for the user network namespace by default, although this can be overridden with the `K3S_ROOTLESS_CIDR` environment variable. Slirp4netns handles routing traffic between the actual host network namespace, and the user network namespace. It also handles mapping unprivileged ports on the host to privileged ports in the user network namespace.

The current unsolved challenges probably mostly revolve around how to get the Flannel CNI working through the port mapping; at the moment there is no way to tell either k3s or flannel to communicate with the peer using the 10000-port offset. It would take some engineering work across k3s and flannel to get all that working properly.

**Reply by vsoch** [reply]
Thank you it makes much more sense. The rootless setup is expecting that endpoint to be in this isolated namespace and it is not. I haven't contributed to flannel or the project here, but I'm fairly OK with Go and developing for Kubernetes and containers, so (if you are able) if you wanted to give me high level places to look and things to try, I could probably take a shot. It sounds like we would hit this issue regardless of the tool we tried (e.g., I was also looking at usernetes but haven't tried it yet) so aside from being interesting / and I'd learn a lot, it could be at least a start toward this vision. Over in HPCLand we are very interested in embracing Kubernetes, and while we do have access to clouds, the ultimate goal is to get it working on our current resources (of course with these constraints). Let me know your thoughts and if there is a logical entrypoint for me to help.

**Reply by brandond** [reply]
I would probably try to figure out what bit of it to tackle first. If your most important use case involves a rootless server and rootless agents running on different hosts, I would probably try to figure out how to get vxlan working with the port offset, and through slirp4netns. I have done zero investigation on that, @AkihiroSuda may have some suggestions.

This is all assuming that you can figure out how to coordinate using unique rootless CIDRs across all your nodes using some automation in your provisioning framework.

**Reply by vsoch** [reply]
> If your most important use case involves a rootless server and rootless agents running on different hosts

Yes! That is absolutely the most important bit.

> This is all assuming that you can figure out how to coordinate using unique rootless CIDRs across all your nodes using some automation in your provisioning framework.

We'll have to figure this out as well - I'm good just starting with step 1 and having this come after it.

@AkihiroSuda let me know if you have some pointers before I try to figure out these different pieces - even just links / comments or ideas would be helpful.

**Reply by AkihiroSuda** [reply]
>  I would probably try to figure out how to get vxlan working with the port offset, and through slirp4netns. I have done zero investigation on that, @AkihiroSuda may have some suggestions.

https://github.com/rootless-containers/usernetes/tree/v20230518.0#ip-addresses
https://github.com/rootless-containers/usernetes/blob/58df6ea63cc4a00425b80a088889015eedc96320/docker-compose.yml#L67

In usernetes, each of the nodes has to be explicitly configured with `--cidr=10.0.XXX.0/24`.
k3s should follow this too.

**Reply by vsoch** [reply]
In case anyone can help with the usernetes stuff (reading here) I have a setup I'm working on that (hopefully) will get the docker-compose variant working on networked VMs! https://github.com/rootless-containers/usernetes/issues/281
